src/Microsoft.IdentityModel.Protocols.OpenIdConnect/OpenIdConnectProtocolValidator.cs

```diff
@@ -178,110 +185,222 @@ public TimeSpan NonceLifetime
-        /// <remarks>It is assumed that the IdToken has had basic validation performed.</remarks>
-        public virtual void Validate(OpenIdConnectProtocolValidationContext validationContext)
+        /// <remarks>It is assumed that the IdToken has had basic validation ('aud', 'iss', 'signature', 'lifetime') performed.</remarks>
+        public virtual void ValidateAuthenticationResponse(OpenIdConnectProtocolValidationContext validationContext)
         {
             if (validationContext == null)
             {
                 LogHelper.Throw(string.Format(CultureInfo.InvariantCulture, LogMessages.IDX10000, GetType() + ": validationContext"), typeof(ArgumentNullException), EventLevel.Verbose);
             }
 
-            // this means the 'id_token' was not validated
-            if (validationContext.IdToken == null && validationContext.ProtocolMessage != null && !string.IsNullOrEmpty(validationContext.ProtocolMessage.IdToken))
+            // no 'response' is recieved or 'id_token' in the response is null 
+            if (validationContext.ProtocolMessage == null)
+            {
+                LogHelper.Throw(LogMessages.IDX10333, typeof(OpenIdConnectProtocolException), EventLevel.Error);
+            }
+
+            if (string.IsNullOrEmpty(validationContext.ProtocolMessage.IdToken))
+            {
+                // if 'code' is also not present, then throw.
+                if (string.IsNullOrEmpty(validationContext.ProtocolMessage.Code))
+                {
+                    LogHelper.Throw(LogMessages.IDX10334, typeof(OpenIdConnectProtocolException), EventLevel.Error);
+                }
+                else
+                {
+                    ValidateState(validationContext);
+                }
+                return;
+            }
+
+            if (validationContext.ValidatedIdToken == null)
             {
                 LogHelper.Throw(LogMessages.IDX10331, typeof(OpenIdConnectProtocolException), EventLevel.Error);
             }
 
-            if (validationContext.IdToken != null)
+            // 'refresh_token' should not be returned from 'authorization_endpoint'. http://tools.ietf.org/html/rfc6749#section-4.2.2.
+            if (!string.IsNullOrEmpty(validationContext.ProtocolMessage.RefreshToken))
             {
-                ValidateIdToken(validationContext.IdToken);
-                ValidateCHash(validationContext);
-                ValidateAtHash(validationContext);
-                ValidateNonce(validationContext);
+                LogHelper.Throw(LogMessages.IDX10335, typeof(OpenIdConnectProtocolException), EventLevel.Error);
             }
 
             ValidateState(validationContext);
+            ValidateIdToken(validationContext.ValidatedIdToken, validationContext);
+            ValidateNonce(validationContext.ValidatedIdToken, validationContext);
+            ValidateCHash(validationContext.ValidatedIdToken, validationContext);
```

Comment:
We need to open an issue that there doesn't seem to be any crypto agility for users to plug in their own hashing algorithms. Users can do this when the validate the signature, but not when validating the at or c hashes.